### DISCUSSION TITLE
pass custom canvas into DeckGL

### DISCUSSION THREAD

**Discussion by VincentNevermore** [visgl/deck.gl#7557] [Q&A]
Hi All, I am currently working on a task that I need to share webGL context to different DeckGL components. (e.g Once a deckGL component is generated, I want to "copy and paste" it and apply to another DeckGL component that will be exactly the same compare to the other one). I noticed that there is two props: canvas and gl introduced in the Document. https://deck.gl/docs/api-reference/core/deck#gl 

So to start with I want to pass my custom canvas into the DeckGL component and to see if it works nor not using the canvas parameter.
However I am getting the folloing error and I have created a simple code to describe this issue:

    

    const canvasRef = React.useRef<HTMLCanvasElement>(null);
    const canvas = document.createElement("canvas");

    React.useEffect(() => {
        if (canvasRef.current) {
            const context = canvasRef.current?.getContext("2d");
            if (context) {
                context.beginPath();
                context.arc(50, 50, 50, 0, 2 * Math.PI);
                context.fill();
            }
        }
    }, []);

    return (
        <>
            <div className={useStyles().main}>
              ###   **{/* <canvas ref={canvasRef} height={500} width={500} /> */}**   ------ LINE
                <DeckGL
                    // @ts-expect-error this prop doesn't exists directly on DeckGL, but on Deck.Context
                    canvas={canvas}
                />
            </div>
        </>
    );

if I uncomment the LINE as mentioned in top then there will be solid circle to be drawn. However if I pass this into deckGl directly, i got a blank page. 

![image](https://user-images.githubusercontent.com/39239025/210595320-b38cd1c3-067f-484d-aaab-8eb99e818a3f.png)

does anyone know what is going on about it ? thanks

**Comment by Pessimistress** [comment]
The React component will always call `new Deck` with the canvas instance that itself creates. In short if you want to use that prop then don't use the React component.

> share webGL context to different DeckGL components

Not sure what you are trying to achieve there. Each Deck instance will have its own redraw loop and they will erase the output of each other.

**Reply by VincentNevermore** [reply]
Thanks for the answer. For this part:

share webGL context to different DeckGL components

Let me clarify and make it more clear about what I am trying to do. The idea is that we have another component that is on top of this DeckGL react component. What I mean share webGL context here is that we want to expose this GL context that has been used and try to share it with other components.(i.e kind of find a way to avoid 16 max live context limit)  I guess from what you mentioned is that this GL context is not able to be shared because of it will erased at the deck of the loop ? Sorry for the confusion.

**Reply by Pessimistress** [reply]
If you are sharing the context with another (not DeckGL) component, then it makes more sense. Still hitting the context limit sounds extreme. Are you making multiple plots side-by-side? If so, you should checkout the multiple views system instead of creating one Deck for each view.

Sharing context is quite tricky because by default, Deck will wipe the context before each redraw. When you pass the context to another WebGL library, you also run into the issue that the context state changes unexpectedly when it is handed back. We do share the gl context with Mapbox and google maps in the respective integration modules. You can look at the source code if you need some references. Overall I do not recommend it if you can avoid it at all. There are a lot of edge cases that make it a game of forever wack-a-mole whenever either library updates.

**Comment by hkfb** [comment]
Maybe this could work: https://github.com/greggman/virtual-webgl

It allows creating several canvases without exhausting the WebGL context limit.

**Comment by hkfb** [comment]
Would it be feasible to do something like this with deck.gl: https://threejs.org/examples/webgl_multiple_elements.html

I believe it uses webgl scissor test to limit drawing to specific areas and overlaying on top of the document.

**Reply by Pessimistress** [reply]
You can probably use multiple views to achieve this. You will need to recalculate the position of each view on scroll. I don't believe the current view system allows you to define a rect that is outside the canvas, so you will need to crop/remove some of them as they move out of the page.
